patches/server/0949-Paper-Plugins.patch

```diff
@@ -0,0 +1,6590 @@
++
++        try {
++            FILE_PROVIDER_SOURCE.registerProviders(runtimePluginEntrypointHandler, path);
++        } catch (IllegalArgumentException exception) {
++            return null; // Return null when the plugin file is not valid / plugin type is unknown
++        } catch (PluginGraphCycleException exception) {
++            throw new InvalidPluginException("Cannot import plugin that causes cyclic dependencies!");
++        } catch (SerializationException |
++                 InvalidDescriptionException ex) { // The spigot implementation wraps it in an invalid plugin exception
++            throw new InvalidPluginException(ex);
++        } catch (Exception e) {
++            throw new InvalidPluginException(e);
++        }
++
++        try {
++            runtimePluginEntrypointHandler.enter(Entrypoint.PLUGIN);
++        } catch (Throwable e) {
++            throw new InvalidPluginException(e);
++        }
++
++        return runtimePluginEntrypointHandler.getPluginProviderStorage().getSingleLoaded()
++            .orElseThrow(() -> new InvalidPluginException("Plugin didn't load any plugin providers?"));
++    }
++
++    // The behavior of this is that all errors are logged instead of being thrown
++    public @NotNull Plugin[] loadPlugins(@NotNull Path directory) {
++        Preconditions.checkArgument(Files.isDirectory(directory), "Directory must be a directory"); // Avoid creating a directory if it doesn't exist
++
++        RuntimePluginEntrypointHandler<MultiRuntimePluginProviderStorage> runtimePluginEntrypointHandler = new RuntimePluginEntrypointHandler<>(new MultiRuntimePluginProviderStorage());
++        try {
++            DIRECTORY_PROVIDER_SOURCE.registerProviders(runtimePluginEntrypointHandler, directory);
++            runtimePluginEntrypointHandler.enter(Entrypoint.PLUGIN);
++        } catch (Exception e) {
++            // This should never happen, any errors that occur in this provider should instead be logged.
++            this.server.getLogger().log(Level.SEVERE, "Unknown error occurred while loading plugins through PluginManager.", e);
++        }
++
++        return runtimePluginEntrypointHandler.getPluginProviderStorage().getLoaded().toArray(new JavaPlugin[0]);
++    }
++
++    // Plugins are disabled in order like this inorder to "rougly" prevent
++    // their dependencies unloading first. But, eh.
++    public void disablePlugins() {
++        Plugin[] plugins = this.getPlugins();
++        for (int i = plugins.length - 1; i >= 0; i--) {
++            this.disablePlugin(plugins[i]);
++        }
++    }
++
++    public void clearPlugins() {
++        synchronized (this) {
++            this.disablePlugins();
++            this.plugins.clear();
++            this.lookupNames.clear();
++        }
++    }
++
++    public synchronized void enablePlugin(@NotNull Plugin plugin) {
++        if (!plugin.isEnabled()) {
```

Comment:
Early return instead of indenting the entire method body when you merge the patches